### ISSUE TITLE
Better test doubles for Xhr

### ISSUE THREAD

**Issue by taymonbeal** [ampproject/amphtml#8020]
As far as my team's aware, the current state of the art for test doubles that mimic `Xhr`-initiated network requests is Sinon stubs on methods like `Xhr.prototype.fetchJson`. These are very fragile; a change to how the code under test consumes the `Xhr` API can break such a test, even if the code's actual behavior with respect to the network is unchanged. A proper test double for `Xhr` should allow the test author to specify network requests in terms of their behavior (e.g., HTTP method, URL, body) and control the corresponding response behavior (i.e., cause the request to yield a provided `Response`, or to fail, at a time of the test author's choosing). The test author should also be able to make assertions about the number and properties of network requests that have been sent as of a given time.

If there's existing tooling that does this, I'd be interested to hear about it. I think Sinon has some kind of proxy server but it doesn't work with native `fetch`. If there isn't, would it be feasible to implement?

**Comment by taymonbeal** [comment]
Here's an idea for an API design for this. What do you think? Adding @rsimha-amp.

```javascript
/**
 * Globally mocks the `Xhr` class implementation to make mock network requests
 * instead of real ones. After this method returns, the next time a network
 * request would be initiated via `Xhr`, instead the `request` callback of the
 * first `MockEntry` in `entries` is called. This function can contain
 * assertions on the characteristic of that request. Assuming `request` returns
 * successfully, the `Xhr` method returns as though the server had sent a
 * response as described by the `response` of the current `MockEntry`. A
 * `Promise` that resolves or rejects at a later time may be used to simulate a
 * delay in the server's response. The next `Xhr`-initiated network request uses
 * the next `MockEntry` in `entries`, and so forth. If no more `entries` remain,
 * an `AssertionError` is thrown.
 *
 * @param {string} origin the security origin that requests will be considered
 *     to have come from. Used to mock CORS behavior.
 * @param {!Array<!MockEntry>} entries the entries for all requests that are
 *     expected to be made in this test
 * @throws {!Error} if `Xhr` has already been mocked with this function
 */
export function mockXhr(origin, entries) { /* ... */ }

/**
 * If `Xhr` is not currently mocked with `mockXhr`, does nothing. If it is
 * currently mocked, asserts that all `entries` that were passed to `mockXhr`
 * have been used. Also restores `Xhr` to its unmocked state, regardless of
 * whether the assertion succeeds. Test suites that use `mockXhr` should pass
 * this function to `afterEach`.
 *
 * @throws {!AssertionError} if unused `entries` remain from `mockXhr`
 */
export function unmockXhr() { /* ... */ }

/**
 * @typedef {{
 *   request: function(!MockRequest),
 *   response: !Promise<!MockResponse>,
 * }}
 */
export let MockEntry;

/**
 * @typedef {{
 *   method: string,
 *   url: string,
 *   headers: !Object<string, string>,
 *   body: string,
 * }}
 */
export let MockRequest;

/**
 * @typedef {{
 *   status: number,
 *   headers: !Object<string, string>,
 *   body: string,
 * }}
 */
export let MockResponse;
```

An example of how this would be used in tests:

```javascript
import {mockXhr, unmockXhr} from './mock-xhr';
import {xhrFor} from '../src/services';

describe('mock-xhr', () => {

  afterEach(unmockXhr);

  it('should mock fetchJson', () => {
    mockXhr('https://example.com', [{
              request(req) {
                expect(req.method).to.equal('GET');
                expect(req.url).to.equal('https://example.com/endpoint.json');
                expect(req.headers['Accept']).to.equal('application/json');
                expect(req.body).to.be.empty;
              },
              response: Promise.resolve({
                status: 200,
                headers: {'Content-Type': 'application/json'},
                body: '{"a":1,"b":2}',
              }),
            }]);
    return expect(xhrFor(window).fetchJson('https://example.com/endpoint.json'))
        .to.eventually.deep.equal({'a': 1, 'b': 2});
  });

});
```

**Comment by erwinmombay** [comment]
adding @dvoytenko as well since i know he's looked/thought about this

**Comment by dvoytenko** [comment]
Yeah, I can see this. Probably would have to be only allowed via `describes` b/c it'd need access to pre-created `sandbox`.

**Comment by taymonbeal** [comment]
Why would we need a Sinon sandbox? Those aren't any kind of magic, they just monkeypatch the object under test, and this could do the same directly.

Would it be okay for me to put together a pull request for this?

**Comment by rsimha-amp** [comment]
@taymonbeal Can this be closed now that #11001 has been merged?

**Comment by taymonbeal** [comment]
Yes. Closing as duplicate of #10477.
